File: code-signal-arcade-universe/clojure/minesweeper.clj
(defn neighbours [matrix [x y]]
  (for [dx [-1 0 1]
        dy [-1 0 1]
        :when (not= 0 dx dy)]
    (get-in matrix [(+ x dx) (+ y dy)])))

(defn no-mine [matrix pos]
  (->> (neighbours matrix pos)
       (filter true?)
       count))

(defn minesweeper [matrix]
  (let [x (count matrix)
        y (count (first matrix))
        z (for [i (range x)
                j (range y)]
            (no-mine matrix [i j]))]
    (partition y z)))


(comment "
(defn neighbours [matrix [x y]]
  (for [dx [-1 0 1]
        dy [-1 0 1]
        :when (not= 0 dx dy)]
    (get-in matrix [(+ x dx) (+ y dy)])))

(defn no-mine [matrix pos]
  (->> (neighbours matrix pos)
       (filter true?)
       count))

(defn minesweeper [matrix]
  (let [x (count matrix)
        y (count (first matrix))
        z (for [i (range x)
                j (range y)]
            (no-mine matrix [i j]))]
    (partition y z)))
")

File: code-signal-arcade-universe/python3/minesweeper.py
def minesweeper(m):
    r = []
    for i in range(len(m)):
        r.append([])
        for j in range(len(m[0])):
            n = -m[i][j]
            for x in [-1, 0, 1]:
                for y in [-1, 0, 1]:
                    if 0 <= i + x < len(m) and 0 <= j + y < len(m[0]):
                        n += m[i+x][j+y]
            r[i].append(n)
    return r

"""
In the popular Minesweeper game you have a board with some mines and those cells that don't contain a mine have a number in it that indicates the total number of mines in the neighboring cells. Starting off with some arrangement of mines we want to create a Minesweeper game setup.

Example

For

matrix = [[true, false, false],
          [false, true, false],
          [false, false, false]]
the output should be

minesweeper(matrix) = [[1, 2, 1],
                       [2, 1, 1],
                       [1, 1, 1]]

Input/Output

[execution time limit] 4 seconds (py3)

[input] array.array.boolean matrix

A non-empty rectangular matrix consisting of boolean values - true if the corresponding cell contains a mine, false otherwise.

Guaranteed constraints:
2 ≤ matrix.length ≤ 100,
2 ≤ matrix[0].length ≤ 100.

[output] array.array.integer

Rectangular matrix of the same size as matrix each cell of which contains an integer equal to the number of mines in the neighboring cells. Two cells are called neighboring if they share at least one corner.
"""
